### ISSUE TITLE
New Annotations (enhancement request)

### ISSUE THREAD

**Issue by si-robertson** [google/closure-compiler#824]
Hi guys,

I'm currently working on a tool that converts Web IDL definitions to externs (a brief discussion can be found in the [closure-compiler-discuss](https://groups.google.com/forum/#!topic/closure-compiler-discuss/0vYO7JsRLj4) group) and I believe a couple of new annotations for externs would be beneficial for developers and compile-time error checking.

These annotations would allow the compiler to determine if an interface or interface member is available within registered global scopes (e.g. Window, Worker) and report an error when an API is not available within the global scope that a script is being compiled for.

`@scope {interface-identifiers...}` tells the compiler that an interface or interface member is only available within one or more global scopes. If the attribute is not specified then it implicitly defaults to `Window`. If any of the `interface-identifiers` have not been registered the compiler should flag the annotation as an error. All members of an interface implicitly inherit the interface's `@scope` annotation.

````JavaScript
/**
 * @scope {Window} // the default if unspecified
 * @constructor
 */
function Audio() {} // not available in workers
````

`@global` tells the compiler to register an interface as a global scope.

````JavaScript
/**
 * @global
 * @constructor
 */
function Window() {}
````

Those annotations would introduce similar behaviour to the Web IDL `Exposed` and `Global` extended attributes which limit the availability of APIs within global scopes.

Developers could choose the global scope they wish to compile their scripts for by using a compiler option. For example, a script intended to be used within a dedicated worker could be compiled with the `--global_scope=DedicatedWorker` option (or something along those lines).

----
Pinging: @ChadKillingsworth

**Comment by ChadKillingsworth** [comment]
I think this would require major changes in the compiler. The compiler is built with only a single global scope in mind.

As for implementation, I was hoping the `--jscontext` flag you proposed would simply load distinct sets of externs. If an extern is valid in more than one set, we could just duplicate it. This shouldn't be a problem when they are auto-generated.

We would of course need a base set that was always included (es3, es5, es6, etc). I'm fine with manually maintaining the core JS externs. It's the web interface ones that change the most.

On that same note, you might consider having an extern file for each IDL file. A one-to-one mapping would be very convenient for maintenance

**Comment by si-robertson** [comment]
Splitting up the externs for each global scope won't be a problem, in fact that's probably the best way to handle things. What I will do is generate a directory structure for the global scopes and populate it with individual extern files (e.g. one extern per interface).

I'm going to close this issue because the requested annotations won't be needed now.

As far as various ES flavors go that's going to be difficult to automate because IDL files don't contain that information. The only way I can think of dealing with that is with a table that maps ES versions to the APIs, but that will be a monumental manual task and will need to be kept up to date.
